src/api.rs

```diff
@@ -196,41 +214,54 @@ impl API {
       .map(|drop| drop.drinkBalance)
   }
 
-  pub fn get_token(self: &mut API) -> Result<String, Box<dyn std::error::Error>> {
-    return match &self.token {
+  fn take_token(&self, token: &mut Option<String>) -> Result<String, APIError> {
+    match token {
       Some(token) => Ok(token.to_string()),
       None => {
         let response = Request::get("https://sso.csh.rit.edu/auth/realms/csh/protocol/openid-connect/auth?client_id=clidrink&redirect_uri=drink%3A%2F%2Fcallback&response_type=token%20id_token&scope=openid%20profile%20drink_balance&state=&nonce=")
           .authentication(Authentication::negotiate())
-          .body(())?.send()?;
+          .body(()).map_err(APIError::HTTPError)?.send().map_err(APIError::IsahcError)?;
         let location = match response.headers().get("Location") {
           Some(location) => location,
           None => {
             API::login();
-            return self.get_token();
+            return self.take_token(token);
           }
         };
-        let url = Url::parse(&location.to_str()?.replace('#', "?"))?;
+        let url = Url::parse(
+          &location
+            .to_str()
+            .map_err(|_| APIError::BadFormat)?
+            .replace('#', "?"),
+        )
+        .map_err(|_| APIError::BadFormat)?;
 
         for (key, value) in url.query_pairs() {
           if key == "access_token" {
             let value = format!("Bearer {}", value);
-            self.token = Some(value.clone());
+            *token = Some(value.clone());
             return Ok(value);
           }
         }
-        return Err(Box::new(APIError::BadFormat));
+        Err(APIError::BadFormat)
       }
-    };
+    }
+  }
+
+  pub fn get_token(&self) -> Result<String, APIError> {
+    let mut token = self.token.lock().unwrap();
+    self.take_token(token.deref_mut())
   }
 
   fn login() {
     // Get credentials
-    let username: Option<String> = std::env::var("CLINK_USERNAME")
-      .map(Some)
-      .unwrap_or_else(|_| get_current_username().and_then(|it| it.into_string().ok()));
-
-    let username: String = username.unwrap();
+    let username: String = match std::env::var("CLINK_USERNAME") {
```

Comment:
Ooh I love nested matches like this that can be streamlined with combinators. Something like this should work

```rust
let username: String = std::env::var("CLINK_USERNAME")
    .ok()
    .or_else(|| get_current_username().and_then(|it| it.into_string() /* might need .ok() here */))
    .or_else(|| std::env::var("USER").ok())
    .expect("Couldn't determine username");
```